viewing_party/party.py

```diff
@@ -1,23 +1,155 @@
 # ------------- WAVE 1 --------------------
 
+import re
+import string
+from venv import create
+from xml.dom import IndexSizeErr
+
+
+
 def create_movie(title, genre, rating):
-    pass
+    movies = {}
+
+    if title != None and genre != None and rating != None:
+            movies.update({"title": title })
+            movies.update({"genre": genre})
+            movies.update({"rating": float(rating) })
+    else:
+        return None
+    
+    return movies
+
+def add_to_watched(user_data, movies):
+    
+    user_data["watched"].append(movies)
+
+    return user_data
+
+
+def add_to_watchlist(user_data, movies):
+
+    user_data["watchlist"].append(movies)
+
+    return user_data
+
+#adds movie to watched list
+def watch_movie (user_data, movie_to_watch):
+    
+        for movie in user_data["watchlist"]:
+            if movie["title"] == movie_to_watch:
+                user_data["watched"].append(movie)
+                user_data["watchlist"].remove(movie)
+        return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    all_watched_ratings = []
+    for movie in user_data["watched"]:
+        all_watched_ratings.append(movie["rating"])
+
+    if len(all_watched_ratings) > 0:
+        return sum(all_watched_ratings) / len(all_watched_ratings)
+    else:
+        return 0
+
+#genre is a string
+def get_most_watched_genre(user_data):
+
+    if user_data["watched"] != []:
+        try:
+
+            most_watched = {}
+            description = []
+            for movie in user_data["watched"]:
+                if movie['genre']:
+                    description.append(movie['genre'])
+            for genre in description:
+                if genre in most_watched:
+                    most_watched[genre] += 1
+                else:
+                    most_watched[genre] = 1
+            most_popular = max(most_watched.values())
+            for keys, values in most_watched.items():
+                if values == most_popular:
+                    return keys
+                return user_data
+        except None:
+            return  None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    unique_movies = []
+    for movie in user_data["watched"]:
+        watched_flag = False
+        for friend in user_data["friends"]:
+            if watched_flag:
+                break
+            for friend_movie in friend["watched"]:
+                if movie == friend_movie:
+                    watched_flag = True
+        if not watched_flag:

```

Comment:
This all boils down to "if the movie is not in any of the friends lists:" but notice that it's a little complicated to read. Some alternatives to consider would be to calculate a combined list of all the movies the friends have watched before we do this check. If that were packed into a `friends_watched_list`, then all of the code here could be replaced with
```python
    if movie not in friends_watched_list:
```
which mirrors the text description above very well.

We could go a step further an instead make a set of the titles of the movies the friends have watched (we can't put a whole movie in a set, since it's not immutable). Then to see whether to add a movie from the user's list, we check that its title is not in the friends watched title set. Checking set membership is much more efficient than checking whether a value is in a list.